sysdep/unix/krt.c

```diff
@@ -723,10 +730,27 @@ krt_got_route(struct krt_proto *p, rte *e)
 
       if (!new)
 	verdict = KRF_DELETE;
-      else if ((net->n.flags & KRF_SYNC_ERROR) || !krt_same_dest(e, new))
-	verdict = KRF_UPDATE;
       else
-	verdict = KRF_SEEN;
+      {
+        /* Replace interface with tunnel if there is set tunnel attribute.
+         * EA_KRT_TUNNEL - calico-bird specific
+         */
+        eattr *ea = ea_find(tmpa, EA_KRT_TUNNEL);
+        struct iface* i = NULL;
+        int check_iface = 0;
+        if (ea && (i = if_find_by_name(ea->u.ptr->data)))
+        {
+          /* As long as we change the interface, we must also check interfaces
+           * equality.
+           */
+          new->attrs->iface = i;
+          check_iface = 1;
+        }
+        if ((net->n.flags & KRF_SYNC_ERROR) || !krt_same_dest(e, new, check_iface))
+          verdict = KRF_UPDATE;
+        else
+          verdict = KRF_SEEN;
+      }
```

Comment:
I think this commit could be simpler and safer if
- the iface comparison logic is implemented right here, instead of in krt_same_dest
- new->attrs->iface is not changed.

It would be simpler because then - I think - there would only be a code change in this function, and no change would be needed in nl_send_route, or in krt_same_dest.  Then the logic needed here is just:

- If there's an EA_KRT_TUNNEL attribute, and the specified iface != e->attrs->iface, set verdict = KRF_UPDATE
- Else the pre-existing logic.

Have I missed anything?  If not, I would prefer that.